# <a id='toc1_'></a>[Normalizer Performance Analysis](#toc0_)

This notebook contains an analysis of the normalizer performance on the CIViC, MOA, and Clinvar data

## <a id='toc1_1_'></a>[Import relevant packages](#toc0_)

In [1]:
import pandas as pd
import numpy as np
import json
import re
import plotly.graph_objects as go
from enum import Enum
from enum import IntEnum

## <a id='toc1_2_'></a>[Dictionaries to map variants to categories and record category counts](#toc0_)

Bin variants to categories.

For variants with multiple associated types:  If the 2+ types have a subset relationship (eg frameshift; frameshift truncation), they are assigned to categories consistent with the superset type (frameshift).  If the types are disjoint (eg: Transcript Variant; Loss of Function Variant), they are assigned with the category most closely associated with the assayed data (Transcript Variant).  This assignment is done in the civic_category_bins dictionary.

In [2]:
civic_category_bins = {
    "Delins":"Sequence Variants",
    "Direct Tandem Duplication":"Sequence Variants",
    "Disruptive Inframe Deletion":"Sequence Variants",
    "Disruptive Inframe Insertion":"Sequence Variants",
    "Coding Sequence Variant":"Sequence Variants",
    "Conservative Inframe Deletion":"Sequence Variants",
    "Copy Number Variants":"Copy Number Variants",
    "Frameshift":"Sequence Variants",
    "Frameshift Truncation":"Sequence Variants",
    "Frameshift Variant":"Sequence Variants",
    "Frameshift Variant;Minus 1 Frameshift Variant":"Sequence Variants",
    "Inframe Deletion":"Sequence Variants",
    "Inframe Indel":"Sequence Variants",
    "Inframe Insertion":"Sequence Variants",
    "Intron Variant":"Region-Defined Variants",
    "Minus 1 Frameshift Variant":"Sequence Variants",
    "Minus 2 Frameshift Variant":"Sequence Variants",
    "Missense Variant":"Sequence Variants",
    "Non Conservative Missense Variant":"Sequence Variants",
    "Plus 1 Frameshift Variant":"Sequence Variants",
    "Region-Defined Variant":"Region-Defined Variants",
    "Regulatory Region Variant":"Region-Defined Variants",
    "Sequence Variants":"Sequence Variants",
    "Splice Acceptor Variant":"Region-Defined Variants",
    "Splice Donor Region Variant":"Region-Defined Variants",
    "Splice Donor Variant":"Region-Defined Variants",
    "Splicing Variant":"Other Variants",
    "Start Lost":"Sequence Variants",
    "Stop Gained":"Sequence Variants",
    "Stop Lost":"Sequence Variants",
    "Synonymous Variant":"Sequence Variants",
    "Transcript Amplification":"Copy Number Variants",
    "Transcript Fusion":"Fusion Variants",
    "3 Prime UTR Variant":"Region-Defined Variants",
    "Amino Acid Deletion;Inframe Deletion":"Sequence Variants",
    "Frameshift Truncation;Minus 2 Frameshift Variant":"Sequence Variants",
    "Frameshift Truncation;Plus 2 Frameshift Variant":"Sequence Variants",
    "Frameshift Variant;Delins":"Sequence Variants",
    "Inframe Insertion;Delins":"Sequence Variants",
    "Inframe Insertion;Inframe Deletion;Delins":"Sequence Variants",
    "Inframe Variant;Inframe Insertion;Inframe Deletion;Delins ":"Sequence Variants",
    "Minus 1 Frameshift Variant;Frameshift Truncation":"Sequence Variants",
    "Plus 1 Frameshift Variant;Frameshift Elongation":"Sequence Variants",
    "Plus 1 Frameshift Variant;Frameshift Truncation":"Sequence Variants",
    "Missense Variant;Gain Of Function Variant":"Sequence Variants", 
    "Missense Variant;Loss Of Function Variant":"Sequence Variants", 
    "Missense Variant;Loss Of Heterozygosity":"Sequence Variants", 
    "Missense Variant;Polymorphic Sequence Variant":"Sequence Variants", 
    "Missense Variant;Snp":"Sequence Variants", 
    "Missense Variant;Transcript Fusion":"Sequence Variants",
    "Stop Gained;Loss Of Function Variant":"Sequence Variants",
    "Stop Lost;Inframe Deletion":"Sequence Variants"
}



moa_category_bins = {
    "Copy Number Variants": "Copy Number Variants",
    "Expression Variants": "Expression Variants",
    "Other Variants": "Other Variants",
    "Rearrangement Variants": "Rearrangement Variants",
    "Sequence Variants": "Sequence Variants"
}



clinvar_category_bins = {
    "Complex":"Other Variants",
    "CompoundHeterozygote":"Genotype Variants",
    "Deletion":"Sequence Variants",
    "Diplotype":"Genotype Variants",
    "Distinct chromosomes":"Rearrangement Variants",
    "Duplication":"Sequence Variants",
    "Haplotype":"Sequence Variants",
    "Haplotype, single variant":"Sequence Variants",
    "Indel":"Sequence Variants",
    "Insertion":"Sequence Variants",
    "Inversion":"Sequence Variants",
    "Microsatellite":"Sequence Variants",
    "Phase unknown":"Other Variants",
    "Tandem duplication":"Sequence Variants",
    "Translocation":"Rearrangement Variants",
    "Variation":"Other Variants",
    "copy number gain":"Copy Number Variants",
    "copy number loss":"Copy Number Variants",
    "fusion":"Fusion Variants",
    "protein only":"Sequence Variants",
    "single nucleotide variant":"Sequence Variants"
}

These variables flag the fields in the dictionary item values below.  In category_counts, each entry is a list of integer values, representing, in order, the number of tokens normalized of that variant, the number ostensibly supported but unable to be normalized, the number of tokens that are not supported, and the total number of tokens.

In [3]:
class Fields(IntEnum):
    """Create IntEnum for count fields in the category_counts dict."""
    NORMALIZED_COUNT = 0
    UNABLE_TO_NORMALIZE_COUNT = 1
    UNSUPPORTED_COUNT = 2
    TOTAL_COUNT = 3
    PERCENT_NORMALIZED = 4

In [4]:
category_counts = {
    "Copy Number Variants":[0,0,0,0,0.0],
    "Epigenetic Modification":[0,0,0,0,0.0],
    "Expression Variants":[0,0,0,0,0.0],
    "Fusion Variants":[0,0,0,0,0.0],
    "Gene Function Variants":[0,0,0,0,0.0],
    "Genotype Variants":[0,0,0,0,0.0],
    "Other Variants":[0,0,0,0,0.0],
    "Rearrangement Variants":[0,0,0,0,0.0],
    "Region-Defined Variants":[0,0,0,0,0.0],
    "Sequence Variants":[0,0,0,0,0.0]
}

## <a id='toc1_3_'></a>[CIViC](#toc0_)



In order to score the normalizer's performance on the CIViC data, some cleaning is required.

First we need to read in the data that was ostensibly supported, get rid of variants with multiple type labels, and assign variant types to as  many of the entries as possible that have a "Not provided" value for civic_variant_types.

Read in .csv of normalized variants in CIVIC

In [5]:
civic_normalized_df = pd.read_csv("../civic/variation_analysis/able_to_normalize_queries.csv",sep = "\t")
civic_normalized_df.head()

variant_id                                 query query_type  \
0        2489  NC_000003.11:g.10191648_10191649insC    genomic   
1        1988            NC_000003.11:g.10191649A>T    genomic   
2        2488                        3-10191647-T-G    genomic   
3        1986            NC_000003.11:g.10191648G>T    genomic   
4        1987            NC_000003.11:g.10191649A>G    genomic   

   variant_accepted civic_variant_types  \
0              True           Stop Lost   
1              True           Stop Lost   
2              True           Stop Lost   
3              True           Stop Lost   
4              True           Stop Lost   

                                      vrs_id succeeded_endpoint  
0  ga4gh:VA.A34ZoIhq4xBuQbcE3bkj29n6diS6RzLB          normalize  
1  ga4gh:VA.JcEpDvhUtgDWU4A-bxqLUuczBNb8QqRf          normalize  
2  ga4gh:VA.7nGd8dgHbqtxMHk_rLxrB6_IMAzJ8XnH          normalize  
3  ga4gh:VA.AmLtooLEvgdnEHD5YVWk6u1e2XBe7FiP          normalize  
4  ga4gh:VA.KIz00usFWEmJHNyqmVL61obfgfRPgOIa          normalize

Prune columns and add new column to flag as normalized.

In [6]:
pruned_civic_normalized_df = civic_normalized_df[["variant_id","query","query_type","civic_variant_types"]]
pruned_civic_normalized_df.insert(4,"normalization_status","normalized")
pruned_civic_normalized_df.head()

variant_id                                 query query_type  \
0        2489  NC_000003.11:g.10191648_10191649insC    genomic   
1        1988            NC_000003.11:g.10191649A>T    genomic   
2        2488                        3-10191647-T-G    genomic   
3        1986            NC_000003.11:g.10191648G>T    genomic   
4        1987            NC_000003.11:g.10191649A>G    genomic   

  civic_variant_types normalization_status  
0           Stop Lost           normalized  
1           Stop Lost           normalized  
2           Stop Lost           normalized  
3           Stop Lost           normalized  
4           Stop Lost           normalized

Repeat process with the variants that were unable to be normalized.

In [7]:
civic_not_normalized_df = pd.read_csv("../civic/variation_analysis/unable_to_normalize_queries.csv",sep = "\t")
civic_not_normalized_df.shape

(80, 8)

In [8]:
pruned_civic_not_normalized_df = civic_not_normalized_df[["variant_id","query","query_type","civic_variant_types"]]
pruned_civic_not_normalized_df.insert(4,"normalization_status","not_normalized")
pruned_civic_not_normalized_df.head()

variant_id                    query query_type        civic_variant_types  \
0         748               MLH1 *757L    protein                  Stop Lost   
1        3718                 AR A748V    protein               Not provided   
2        3725                 AR A765T    protein               Not provided   
3        4485  ERBB2 A775_G776ins YVMA    protein               Not provided   
4         248               TERT C228T    protein  Regulatory Region Variant   

  normalization_status  
0       not_normalized  
1       not_normalized  
2       not_normalized  
3       not_normalized  
4       not_normalized

Merge these dfs

In [9]:
frames = [pruned_civic_normalized_df, pruned_civic_not_normalized_df]
civic_supported_df = pd.concat(frames)
civic_supported_df.shape

(1956, 5)

Making all queries in all caps to make it easier to account of untyped variants later on.

In [10]:
civic_supported_df["query"] = civic_supported_df["query"].apply(str.upper)

Checking variant types.  The single largest types is "Not provided".  
Most of these look like amino acid substitutions.
Defining a regex to detect these variants and assign "Missense Variant" type to these variants.

In [11]:
civic_supported_df["civic_variant_types"].value_counts(dropna=False)

civic_variant_types
Not provided                                                 816
Missense Variant                                             815
Stop Gained                                                   60
Frameshift Truncation                                         36
Transcript Amplification                                      36
Inframe Deletion                                              35
Minus 1 Frameshift Variant;Frameshift Truncation              21
Inframe Insertion                                             17
Synonymous Variant                                            13
Splice Donor Variant                                          13
Splice Acceptor Variant                                        9
Missense Variant;Gain Of Function Variant                      9
Frameshift Variant                                             7
Stop Lost                                                      6
Conservative Inframe Deletion                                  6
Delin

If a variant does not have an assigned variant type in civic, if it is a protein query and the query matches a regex pattern associated with variant substitutions (such as "PTEN A126D"), then I am re-classifying them as a "Missense Variant" instead.

In [12]:
civic_supported_df["variant flag"] = civic_supported_df["query"].apply(lambda x: bool(re.match("\S+\s+[A-Z]+\d+[A-Z|*]", x)))
civic_supported_df["civic_variant_types"] = np.where((civic_supported_df["query_type"] == "protein") & (civic_supported_df["civic_variant_types"] == "Not provided") & (civic_supported_df["variant flag"]), "Missense Variant", civic_supported_df["civic_variant_types"])

Doing so reduced the 816 untyped variants down to 70.
Checking the remaining weird variants.

In [13]:
untyped_variants = civic_supported_df[civic_supported_df["civic_variant_types"] == "Not provided"]
untyped_variants.head(20)

variant_id                      query query_type civic_variant_types  \
11         3342          KRAS A11_G12INSGA    protein        Not provided   
65         4484     ERBB2 A775_G776INSIVMA    protein        Not provided   
66         2658     ERBB2 A775_G776INSYVMA    protein        Not provided   
67         4483     ERBB2 A775_G776INSYVMA    protein        Not provided   
74         3751  ARHGAP35 A865_L870DELINSV    protein        Not provided   
82         2655          MYB AMPLIFICATION    protein        Not provided   
110        1261         MDM2 AMPLIFICATION    protein        Not provided   
113        1276          SMO AMPLIFICATION    protein        Not provided   
116        1684        PSMD4 AMPLIFICATION    protein        Not provided   
118        2205         FLT4 AMPLIFICATION    protein        Not provided   
119        2240         TLK2 AMPLIFICATION    protein        Not provided   
120        2397         CRKL AMPLIFICATION    protein        Not provided   
121        2550         NRAS AMPLIFICATION    protein        Not provided   
122        2564          MYC AMPLIFICATION    protein        Not provided   
123        2598         IRS2 AMPLIFICATION    protein        Not provided   
124        2619         FLT3 AMPLIFICATION    protein        Not provided   
125        2637          KDR AMPLIFICATION    protein        Not provided   
126        2746        CCND2 AMPLIFICATION    protein        Not provided   
127        2747        CCND3 AMPLIFICATION    protein        Not provided   
128        2844         MITF AMPLIFICATION    protein        Not provided   

    normalization_status  variant flag  
11            normalized         False  
65            normalized         False  
66            normalized         False  
67            normalized         False  
74            normalized         False  
82            normalized         False  
110           normalized         False  
113           normalized         False  
116           normalized         False  
118           normalized         False  
119           normalized         False  
120           normalized         False  
121           normalized         False  
122           normalized         False  
123           normalized         False  
124           normalized         False  
125           normalized         False  
126           normalized         False  
127           normalized         False  
128           normalized         False

Reassigning variants marked as {gene} Amplification as Transcript Amplification Variants

In [14]:
civic_supported_df["variant flag"] = civic_supported_df["query"].apply(lambda x: bool(re.match("\S+\s+AMPLIFICATION", x)))
civic_supported_df["civic_variant_types"] = np.where((civic_supported_df["query_type"] == "protein") & (civic_supported_df["civic_variant_types"] == "Not provided") & (civic_supported_df["variant flag"]), "Transcript Amplification", civic_supported_df["civic_variant_types"])


Reassigning amino acid insertions, delins, and deletions as "Missense Variant", including a couple of variants that have a random space before or after the sequence operation like "INS"

In [15]:
civic_supported_df["variant flag"] = civic_supported_df["query"].apply(lambda x: bool(re.match("\S+\s+[A-Z]+\d+_+[A-Z]+\d+INS+[A-Z]", x)))
civic_supported_df["civic_variant_types"] = np.where((civic_supported_df["query_type"] == "protein") & (civic_supported_df["civic_variant_types"] == "Not provided") & (civic_supported_df["variant flag"]), "Missense Variant", civic_supported_df["civic_variant_types"])


In [16]:
civic_supported_df["variant flag"] = civic_supported_df["query"].apply(lambda x: bool(re.match("\S+\s+[A-Z]+\d+_+[A-Z]+\d+INS+\s+[A-Z]", x)))
civic_supported_df["civic_variant_types"] = np.where((civic_supported_df["query_type"] == "protein") & (civic_supported_df["civic_variant_types"] == "Not provided") & (civic_supported_df["variant flag"]), "Missense Variant", civic_supported_df["civic_variant_types"])

In [17]:
civic_supported_df["variant flag"] = civic_supported_df["query"].apply(lambda x: bool(re.match("\S+\s+[A-Z]+\d+-+\d+\s+INS+[A-Z]", x)))
civic_supported_df["civic_variant_types"] = np.where((civic_supported_df["query_type"] == "protein") & (civic_supported_df["civic_variant_types"] == "Not provided") & (civic_supported_df["variant flag"]), "Missense Variant", civic_supported_df["civic_variant_types"])

In [18]:
civic_supported_df["variant flag"] = civic_supported_df["query"].apply(lambda x: bool(re.match("\S+\s+[A-Z]+\d+_+[A-Z]+\d+DELINS+[A-Z]", x)))
civic_supported_df["civic_variant_types"] = np.where((civic_supported_df["query_type"] == "protein") & (civic_supported_df["civic_variant_types"] == "Not provided") & (civic_supported_df["variant flag"]), "Missense Variant", civic_supported_df["civic_variant_types"])


In [19]:
civic_supported_df["variant flag"] = civic_supported_df["query"].apply(lambda x: bool(re.match("\S+\s+[A-Z]+\d+_+[A-Z]+\d+DEL", x)))
civic_supported_df["civic_variant_types"] = np.where((civic_supported_df["query_type"] == "protein") & (civic_supported_df["civic_variant_types"] == "Not provided") & (civic_supported_df["variant flag"]), "Missense Variant", civic_supported_df["civic_variant_types"])


And assigning missense types to a handful of remaining variants that are non-standard names for genomic and protein sequence variants

In [20]:
civic_supported_df["variant flag"] = civic_supported_df["query"].apply(lambda x: bool(re.match("\S+\s+P\.+[A-Z]+\d+[A-Z]", x)))
civic_supported_df["civic_variant_types"] = np.where((civic_supported_df["query_type"] == "protein") & (civic_supported_df["civic_variant_types"] == "Not provided") & (civic_supported_df["variant flag"]), "Missense Variant", civic_supported_df["civic_variant_types"])


In [21]:
civic_supported_df["variant flag"] = civic_supported_df["query"].apply(lambda x: bool(re.match("\S+[A-Z]+\-+[A-Z]", x)))
civic_supported_df["civic_variant_types"] = np.where((civic_supported_df["query_type"] == "genomic") & (civic_supported_df["civic_variant_types"] == "Not provided") & (civic_supported_df["variant flag"]), "Missense Variant", civic_supported_df["civic_variant_types"])

In [22]:
civic_supported_df["variant flag"] = civic_supported_df["query"].apply(lambda x: bool(re.match("NC_\d+\.+\d+:[A-Z]+\.+\d+[A-Z]+>+[A-Z]", x)))
civic_supported_df["civic_variant_types"] = np.where((civic_supported_df["query_type"] == "genomic") & (civic_supported_df["civic_variant_types"] == "Not provided") & (civic_supported_df["variant flag"]), "Missense Variant", civic_supported_df["civic_variant_types"])

In [23]:
civic_supported_df["variant flag"] = civic_supported_df["query"].apply(lambda x: bool(re.match("NC_\d+\.+\d+:[A-Z]+\.+\d+_+\d+INS+[A-Z]", x)))
civic_supported_df["civic_variant_types"] = np.where((civic_supported_df["query_type"] == "genomic") & (civic_supported_df["civic_variant_types"] == "Not provided") & (civic_supported_df["variant flag"]), "Missense Variant", civic_supported_df["civic_variant_types"])

This last variant is a unique (to this db) nonstandard nomenclature for just some variant in a particular domain, so it is a region-defined variant.

In [24]:
civic_supported_df["variant flag"] = civic_supported_df["query"].apply(lambda x: bool(re.match("DICER1 RNASE IIIB MUTATION", x)))
civic_supported_df["civic_variant_types"] = np.where((civic_supported_df["query_type"] == "protein") & (civic_supported_df["civic_variant_types"] == "Not provided") & (civic_supported_df["variant flag"]), "Region-Defined Variant", civic_supported_df["civic_variant_types"])

Add category column to CIViC df.

In [25]:
civic_supported_df["category"] = civic_supported_df["civic_variant_types"].map(civic_category_bins)
civic_supported_df.tail()

variant_id                     query query_type civic_variant_types  \
75         953          KIT Y553_K558DEL    protein    Inframe Deletion   
76        1542     KIT Y553_W557DELYEVQW    protein    Inframe Deletion   
77        1548          KIT Y570_L576DEL    protein    Inframe Deletion   
78        1630  FLT3 Y591_V592INSVDFREYE    protein    Missense Variant   
79        3724                  AR Y763C    protein    Missense Variant   

   normalization_status  variant flag           category  
75       not_normalized         False  Sequence Variants  
76       not_normalized         False  Sequence Variants  
77       not_normalized         False  Sequence Variants  
78       not_normalized         False  Sequence Variants  
79       not_normalized         False  Sequence Variants

Split df by normalized/not_normalized flag

In [26]:
civic_normalized_df_cats = civic_supported_df[civic_supported_df["normalization_status"] == "normalized"]
civic_normalized_df_cats

variant_id                                 query query_type  \
0           2489  NC_000003.11:G.10191648_10191649INSC    genomic   
1           1988            NC_000003.11:G.10191649A>T    genomic   
2           2488                        3-10191647-T-G    genomic   
3           1986            NC_000003.11:G.10191648G>T    genomic   
4           1987            NC_000003.11:G.10191649A>G    genomic   
...          ...                                   ...        ...   
1871        3161                        3-10183878-G-A    genomic   
1872         877            NC_000020.11:G.58903752C>T    genomic   
1873         731            NC_000003.11:G.37056036G>A    genomic   
1874        3045                          VHL P.F76DEL    protein   
1875        3310                          HDAC9 P.L33R    protein   

       civic_variant_types normalization_status  variant flag  \
0                Stop Lost           normalized         False   
1                Stop Lost           normalized         False   
2                Stop Lost           normalized         False   
3                Stop Lost           normalized         False   
4                Stop Lost           normalized         False   
...                    ...                  ...           ...   
1871      Missense Variant           normalized         False   
1872    Synonymous Variant           normalized         False   
1873  Splice Donor Variant           normalized         False   
1874      Missense Variant           normalized         False   
1875      Missense Variant           normalized         False   

                     category  
0           Sequence Variants  
1           Sequence Variants  
2           Sequence Variants  
3           Sequence Variants  
4           Sequence Variants  
...                       ...  
1871        Sequence Variants  
1872        Sequence Variants  
1873  Region-Defined Variants  
1874        Sequence Variants  
1875        Sequence Variants  

[1876 rows x 7 columns]

In [27]:
civic_not_normalized_df_cats = civic_supported_df[civic_supported_df["normalization_status"] == "not_normalized"]
civic_not_normalized_df_cats

variant_id                     query query_type  \
0          748                MLH1 *757L    protein   
1         3718                  AR A748V    protein   
2         3725                  AR A765T    protein   
3         4485   ERBB2 A775_G776INS YVMA    protein   
4          248                TERT C228T    protein   
..         ...                       ...        ...   
75         953          KIT Y553_K558DEL    protein   
76        1542     KIT Y553_W557DELYEVQW    protein   
77        1548          KIT Y570_L576DEL    protein   
78        1630  FLT3 Y591_V592INSVDFREYE    protein   
79        3724                  AR Y763C    protein   

          civic_variant_types normalization_status  variant flag  \
0                   Stop Lost       not_normalized         False   
1            Missense Variant       not_normalized         False   
2            Missense Variant       not_normalized         False   
3            Missense Variant       not_normalized         False   
4   Regulatory Region Variant       not_normalized         False   
..                        ...                  ...           ...   
75           Inframe Deletion       not_normalized         False   
76           Inframe Deletion       not_normalized         False   
77           Inframe Deletion       not_normalized         False   
78           Missense Variant       not_normalized         False   
79           Missense Variant       not_normalized         False   

                   category  
0         Sequence Variants  
1         Sequence Variants  
2         Sequence Variants  
3         Sequence Variants  
4   Region-Defined Variants  
..                      ...  
75        Sequence Variants  
76        Sequence Variants  
77        Sequence Variants  
78        Sequence Variants  
79        Sequence Variants  

[80 rows x 7 columns]

For each df, Get CIViC Variant counts by category and add to counts dictionary

In [28]:
civic_normalized_category_counts = json.loads(civic_normalized_df_cats["category"].value_counts().to_json())
civic_normalized_category_counts

{'Sequence Variants': 1788,
 'Copy Number Variants': 57,
 'Region-Defined Variants': 28,
 'Fusion Variants': 1,
 'Other Variants': 1}

In [29]:
def add_json_counts(var_category_counts, support_status) -> None:
    """given a JSON of variant categories and counts and whether that dataframe represents normalized, not_normalized, or not_supported variants, adds the counts of variants to dictionary of counts
    
    :param var_category_counts: counts of variants in clinvar with variant type information in JSON format.
    :param support_status: an int flag to indicate if the variants in the dataframe are normalized (0), unable to be normalized (1), or unsupported (2) by the normalizer
    """
    for category, count in var_category_counts.items():
        category_counts[category][support_status] += count
        category_counts[category][Fields.TOTAL_COUNT] += count

In [30]:
add_json_counts(civic_normalized_category_counts, Fields.NORMALIZED_COUNT)
category_counts

{'Copy Number Variants': [57, 0, 0, 57, 0.0],
 'Epigenetic Modification': [0, 0, 0, 0, 0.0],
 'Expression Variants': [0, 0, 0, 0, 0.0],
 'Fusion Variants': [1, 0, 0, 1, 0.0],
 'Gene Function Variants': [0, 0, 0, 0, 0.0],
 'Genotype Variants': [0, 0, 0, 0, 0.0],
 'Other Variants': [1, 0, 0, 1, 0.0],
 'Rearrangement Variants': [0, 0, 0, 0, 0.0],
 'Region-Defined Variants': [28, 0, 0, 28, 0.0],
 'Sequence Variants': [1788, 0, 0, 1788, 0.0]}

In [31]:
civic_not_normalized_category_counts = json.loads(civic_not_normalized_df_cats["category"].value_counts().to_json())
civic_not_normalized_category_counts

{'Sequence Variants': 77, 'Region-Defined Variants': 3}

In [32]:
add_json_counts(civic_not_normalized_category_counts, Fields.UNABLE_TO_NORMALIZE_COUNT)
category_counts

{'Copy Number Variants': [57, 0, 0, 57, 0.0],
 'Epigenetic Modification': [0, 0, 0, 0, 0.0],
 'Expression Variants': [0, 0, 0, 0, 0.0],
 'Fusion Variants': [1, 0, 0, 1, 0.0],
 'Gene Function Variants': [0, 0, 0, 0, 0.0],
 'Genotype Variants': [0, 0, 0, 0, 0.0],
 'Other Variants': [1, 0, 0, 1, 0.0],
 'Rearrangement Variants': [0, 0, 0, 0, 0.0],
 'Region-Defined Variants': [28, 3, 0, 31, 0.0],
 'Sequence Variants': [1788, 77, 0, 1865, 0.0]}

Read in the csv for unsupported variants.  This data was already mapped to categories in civic_variant_analysis.  Therefore, we only need to import the data and perform the count on the category column.

In [33]:
not_supported_variants = pd.read_csv("../civic/variation_analysis/not_supported_variants.csv",sep = "\t")
print(not_supported_variants.shape)
not_supported_variants.head()

(1563, 6)


variant_id gene_name  variant_name civic_variant_types  \
0        4170       VHL           NaN        Not provided   
1        4417       ALK   FBXO11::ALK        Not provided   
2        4214       VHL           NaN        Not provided   
3        4216       VHL           NaN        Not provided   
4        4278       VHL           NaN        Not provided   

              category  variant_accepted  
0  Transcript Variants             False  
1      Fusion Variants             False  
2  Transcript Variants             False  
3  Transcript Variants             False  
4  Transcript Variants             False

Checking Counts.

In [34]:
not_supported_variants["category"].value_counts()

category
Transcript Variants        366
Fusion Variants            294
Expression Variants        287
Sequence Variants          133
Region Defined Variants    129
Rearrangement Variants     116
Gene Function Variants      91
Other Variants              83
Copy Number Variants        34
Genotype Variants           16
Epigenetic Modification     14
Name: count, dtype: int64

There are two small discrepancies here. First, there is a hyphen missing from "Region-Defined Variants" which will cause a key error.  Second, the variants labelled as "Transcript Variants" here should be binned under "Sequence Variants".  Fixing that now.

In [35]:
not_supported_variants["category"].replace("Region Defined Variants", "Region-Defined Variants", inplace=True)
not_supported_variants["category"].replace("Transcript Variants", "Sequence Variants", inplace=True)

In [36]:
not_supported_variants_category_counts = json.loads(not_supported_variants["category"].value_counts().to_json())
not_supported_variants_category_counts

{'Sequence Variants': 499,
 'Fusion Variants': 294,
 'Expression Variants': 287,
 'Region-Defined Variants': 129,
 'Rearrangement Variants': 116,
 'Gene Function Variants': 91,
 'Other Variants': 83,
 'Copy Number Variants': 34,
 'Genotype Variants': 16,
 'Epigenetic Modification': 14}

In [37]:
add_json_counts(not_supported_variants_category_counts, Fields.UNSUPPORTED_COUNT)
category_counts

{'Copy Number Variants': [57, 0, 34, 91, 0.0],
 'Epigenetic Modification': [0, 0, 14, 14, 0.0],
 'Expression Variants': [0, 0, 287, 287, 0.0],
 'Fusion Variants': [1, 0, 294, 295, 0.0],
 'Gene Function Variants': [0, 0, 91, 91, 0.0],
 'Genotype Variants': [0, 0, 16, 16, 0.0],
 'Other Variants': [1, 0, 83, 84, 0.0],
 'Rearrangement Variants': [0, 0, 116, 116, 0.0],
 'Region-Defined Variants': [28, 3, 129, 160, 0.0],
 'Sequence Variants': [1788, 77, 499, 2364, 0.0]}

## <a id='toc1_4_'></a>[MOA](#toc0_)

Read MOA .csv file for Normalized variants

In [38]:
moa_normalized_df = pd.read_csv("../moa/feature_analysis/able_to_normalize_queries.csv",sep = "\t")
print(moa_normalized_df.shape)
moa_normalized_df.head()

(181, 5)


variant_id         query moa_feature_type           category  \
0          71  ABL1 p.T315I  somatic_variant  Sequence Variants   
1          73  ABL1 p.T315A  somatic_variant  Sequence Variants   
2          75  ABL1 p.F317L  somatic_variant  Sequence Variants   
3          76  ABL1 p.F317V  somatic_variant  Sequence Variants   
4          77  ABL1 p.F317I  somatic_variant  Sequence Variants   

                                      vrs_id  
0  ga4gh:VA.M3CbaYfwomLqvJbdK4w-W7V-zw7LdjGj  
1  ga4gh:VA.Z9Ygies0W47sv7z1LXuMrD8ql6eFv_Fb  
2  ga4gh:VA.KthZ0GvMM8IAq73dQ1Bqh9e8Ys0LKL8p  
3  ga4gh:VA.ZQYG0E49jxi2VTHK3Fmfktzg5KBW0is5  
4  ga4gh:VA.KNp3-VqwztR6vlk9k4x1XhebBhFX3nFe

Get variant counts by category, update variant counts df 

In [39]:
moa_normalized_category_counts = json.loads(moa_normalized_df["category"].value_counts().to_json())
moa_normalized_category_counts

{'Sequence Variants': 149, 'Copy Number Variants': 32}

In [40]:
add_json_counts(moa_normalized_category_counts, Fields.NORMALIZED_COUNT)
category_counts

{'Copy Number Variants': [89, 0, 34, 123, 0.0],
 'Epigenetic Modification': [0, 0, 14, 14, 0.0],
 'Expression Variants': [0, 0, 287, 287, 0.0],
 'Fusion Variants': [1, 0, 294, 295, 0.0],
 'Gene Function Variants': [0, 0, 91, 91, 0.0],
 'Genotype Variants': [0, 0, 16, 16, 0.0],
 'Other Variants': [1, 0, 83, 84, 0.0],
 'Rearrangement Variants': [0, 0, 116, 116, 0.0],
 'Region-Defined Variants': [28, 3, 129, 160, 0.0],
 'Sequence Variants': [1937, 77, 499, 2513, 0.0]}

Repeat same process for variants that were supported but failed to normalize.

In [41]:
moa_not_normalized_df = pd.read_csv("../moa/feature_analysis/unable_to_normalize_queries.csv",sep = "\t")
print(moa_not_normalized_df.shape)
moa_not_normalized_df.head()

(0, 7)


Empty DataFrame
Columns: [variant_id, query, moa_feature_type, category, exception_raised, message, warnings]
Index: []

In [42]:
moa_not_normalized_category_counts = json.loads(moa_not_normalized_df["category"].value_counts().to_json())
moa_not_normalized_category_counts

{}

In [43]:
add_json_counts(moa_not_normalized_category_counts, Fields.UNABLE_TO_NORMALIZE_COUNT)
category_counts

{'Copy Number Variants': [89, 0, 34, 123, 0.0],
 'Epigenetic Modification': [0, 0, 14, 14, 0.0],
 'Expression Variants': [0, 0, 287, 287, 0.0],
 'Fusion Variants': [1, 0, 294, 295, 0.0],
 'Gene Function Variants': [0, 0, 91, 91, 0.0],
 'Genotype Variants': [0, 0, 16, 16, 0.0],
 'Other Variants': [1, 0, 83, 84, 0.0],
 'Rearrangement Variants': [0, 0, 116, 116, 0.0],
 'Region-Defined Variants': [28, 3, 129, 160, 0.0],
 'Sequence Variants': [1937, 77, 499, 2513, 0.0]}

Repeat same process for variants that are unsupported.

In [44]:
moa_not_supported_df = pd.read_csv("../moa/feature_analysis/not_supported_variants.csv",sep = "\t")
print(moa_not_supported_df.shape)
print(moa_not_supported_df.head())
moa_not_supported_df["category"].value_counts(dropna=False)

(249, 4)
   variant_id                        query moa_feature_type  \
0           1             BCR--ABL1 Fusion    rearrangement   
1          12                   ALK Fusion    rearrangement   
2          15                          ALK    rearrangement   
3          18            ALK Translocation    rearrangement   
4          21  BRD4 t(15;19) Translocation    rearrangement   

                 category  
0  Rearrangement Variants  
1  Rearrangement Variants  
2  Rearrangement Variants  
3  Rearrangement Variants  
4  Rearrangement Variants  


category
Sequence Variants         181
Rearrangement Variants     35
Copy Number Variants       17
Expression Variants        11
Other Variants              5
Name: count, dtype: int64

In [45]:
moa_not_supported_category_counts = json.loads(moa_not_supported_df["category"].value_counts().to_json())
moa_not_supported_category_counts

{'Sequence Variants': 181,
 'Rearrangement Variants': 35,
 'Copy Number Variants': 17,
 'Expression Variants': 11,
 'Other Variants': 5}

In [46]:
add_json_counts(moa_not_supported_category_counts, Fields.UNSUPPORTED_COUNT)
category_counts

{'Copy Number Variants': [89, 0, 51, 140, 0.0],
 'Epigenetic Modification': [0, 0, 14, 14, 0.0],
 'Expression Variants': [0, 0, 298, 298, 0.0],
 'Fusion Variants': [1, 0, 294, 295, 0.0],
 'Gene Function Variants': [0, 0, 91, 91, 0.0],
 'Genotype Variants': [0, 0, 16, 16, 0.0],
 'Other Variants': [1, 0, 88, 89, 0.0],
 'Rearrangement Variants': [0, 0, 151, 151, 0.0],
 'Region-Defined Variants': [28, 3, 129, 160, 0.0],
 'Sequence Variants': [1937, 77, 680, 2694, 0.0]}

## <a id='toc1_5_'></a>[ClinVar](#toc0_)

Read in the three clinvar csv files.

In [47]:
clinvar_normalized_df = pd.read_csv("../clinvar/clinvar_variation_analysis_output/variation_type_count_supported_df.csv")
print(clinvar_normalized_df.shape)
clinvar_normalized_df.head(20)

(20, 4)


Unnamed: 0          in.variation_type  \
0            0  single nucleotide variant   
1            1                   Deletion   
2            2                Duplication   
3            3           copy number loss   
4            4             Microsatellite   
5            5           copy number gain   
6            6                   Deletion   
7            7                      Indel   
8            8                  Insertion   
9            9                Duplication   
10          10           copy number loss   
11          11                  Inversion   
12          12           copy number gain   
13          13  single nucleotide variant   
14          14                      Indel   
15          15                  Variation   
16          16                  Insertion   
17          17                   Deletion   
18          18                  Variation   
19          19                Duplication   

                            in.vrs_xform_plan.policy    count  
0                                     Canonical SPDI  1934692  
1                                     Canonical SPDI    92875  
2                                     Canonical SPDI    43185  
3                                Absolute copy count    26857  
4                                     Canonical SPDI    26678  
5                                Absolute copy count    25583  
6   Copy number change (cn loss|del and cn gain|dup)    16249  
7                                     Canonical SPDI    10986  
8                                     Canonical SPDI     8865  
9   Copy number change (cn loss|del and cn gain|dup)     8119  
10  Copy number change (cn loss|del and cn gain|dup)     1524  
11                                    Canonical SPDI     1051  
12  Copy number change (cn loss|del and cn gain|dup)     1003  
13                      Remaining valid hgvs alleles      396  
14                      Remaining valid hgvs alleles      340  
15                                    Canonical SPDI      337  
16                      Remaining valid hgvs alleles      191  
17                               Absolute copy count        3  
18                      Remaining valid hgvs alleles        1  
19                               Absolute copy count        1

In [48]:
clinvar_not_normalized_df = pd.read_csv("../clinvar/clinvar_variation_analysis_output/variation_type_count_supported_not_normalized_df.csv")
print(clinvar_not_normalized_df.shape)
clinvar_not_normalized_df.head(10)

(9, 4)


Unnamed: 0          in.variation_type  \
0           0           copy number loss   
1           1           copy number gain   
2           2                   Deletion   
3           3                Duplication   
4           4           copy number loss   
5           5           copy number gain   
6           6  single nucleotide variant   
7           7                  Insertion   
8           8                  Variation   

                           in.vrs_xform_plan.policy  count  
0                               Absolute copy count    475  
1                               Absolute copy count    344  
2  Copy number change (cn loss|del and cn gain|dup)     89  
3  Copy number change (cn loss|del and cn gain|dup)     72  
4  Copy number change (cn loss|del and cn gain|dup)     31  
5  Copy number change (cn loss|del and cn gain|dup)     17  
6                      Remaining valid hgvs alleles     11  
7                      Remaining valid hgvs alleles      1  
8                      Remaining valid hgvs alleles      1

In [49]:
clinvar_not_supported_df = pd.read_csv("../clinvar/clinvar_variation_analysis_output/variation_type_count_not_supported_df.csv")
print(clinvar_not_supported_df.shape)
clinvar_not_supported_df.head(20)

(37, 4)


Unnamed: 0          in.variation_type  in.vrs_xform_plan.policy  count
0            0           copy number gain       NCBI36 genomic only   2897
1            1           copy number loss       NCBI36 genomic only   1749
2            2                   Deletion  No hgvs or location info   1236
3            3                  Diplotype        Genotype/Haplotype    596
4            4                  Haplotype        Genotype/Haplotype    565
5            5             Microsatellite  Invalid/unsupported hgvs    415
6            6                   Deletion  Invalid/unsupported hgvs    413
7            7                  Insertion  No hgvs or location info    382
8            8                  Insertion  Invalid/unsupported hgvs    306
9            9  single nucleotide variant  No hgvs or location info    287
10          10              Translocation  No hgvs or location info    271
11          11                Duplication  No hgvs or location info    265
12          12       CompoundHeterozygote        Genotype/Haplotype    249
13          13                  Variation  No hgvs or location info    206
14          14                      Indel  No hgvs or location info    138
15          15                Duplication  Invalid/unsupported hgvs    116
16          16               protein only  No hgvs or location info     95
17          17                    Complex  No hgvs or location info     72
18          18                   Deletion       NCBI36 genomic only     71
19          19             Microsatellite  No hgvs or location info     67

Add column and map variant types to categories.

In [50]:
clinvar_normalized_df["category"] = clinvar_normalized_df["in.variation_type"].map(clinvar_category_bins)
clinvar_normalized_df.head(20)

Unnamed: 0          in.variation_type  \
0            0  single nucleotide variant   
1            1                   Deletion   
2            2                Duplication   
3            3           copy number loss   
4            4             Microsatellite   
5            5           copy number gain   
6            6                   Deletion   
7            7                      Indel   
8            8                  Insertion   
9            9                Duplication   
10          10           copy number loss   
11          11                  Inversion   
12          12           copy number gain   
13          13  single nucleotide variant   
14          14                      Indel   
15          15                  Variation   
16          16                  Insertion   
17          17                   Deletion   
18          18                  Variation   
19          19                Duplication   

                            in.vrs_xform_plan.policy    count  \
0                                     Canonical SPDI  1934692   
1                                     Canonical SPDI    92875   
2                                     Canonical SPDI    43185   
3                                Absolute copy count    26857   
4                                     Canonical SPDI    26678   
5                                Absolute copy count    25583   
6   Copy number change (cn loss|del and cn gain|dup)    16249   
7                                     Canonical SPDI    10986   
8                                     Canonical SPDI     8865   
9   Copy number change (cn loss|del and cn gain|dup)     8119   
10  Copy number change (cn loss|del and cn gain|dup)     1524   
11                                    Canonical SPDI     1051   
12  Copy number change (cn loss|del and cn gain|dup)     1003   
13                      Remaining valid hgvs alleles      396   
14                      Remaining valid hgvs alleles      340   
15                                    Canonical SPDI      337   
16                      Remaining valid hgvs alleles      191   
17                               Absolute copy count        3   
18                      Remaining valid hgvs alleles        1   
19                               Absolute copy count        1   

                category  
0      Sequence Variants  
1      Sequence Variants  
2      Sequence Variants  
3   Copy Number Variants  
4      Sequence Variants  
5   Copy Number Variants  
6      Sequence Variants  
7      Sequence Variants  
8      Sequence Variants  
9      Sequence Variants  
10  Copy Number Variants  
11     Sequence Variants  
12  Copy Number Variants  
13     Sequence Variants  
14     Sequence Variants  
15        Other Variants  
16     Sequence Variants  
17     Sequence Variants  
18        Other Variants  
19     Sequence Variants

In [51]:
clinvar_not_normalized_df["category"] = clinvar_not_normalized_df["in.variation_type"].map(clinvar_category_bins)
clinvar_not_normalized_df.head(20)

Unnamed: 0          in.variation_type  \
0           0           copy number loss   
1           1           copy number gain   
2           2                   Deletion   
3           3                Duplication   
4           4           copy number loss   
5           5           copy number gain   
6           6  single nucleotide variant   
7           7                  Insertion   
8           8                  Variation   

                           in.vrs_xform_plan.policy  count  \
0                               Absolute copy count    475   
1                               Absolute copy count    344   
2  Copy number change (cn loss|del and cn gain|dup)     89   
3  Copy number change (cn loss|del and cn gain|dup)     72   
4  Copy number change (cn loss|del and cn gain|dup)     31   
5  Copy number change (cn loss|del and cn gain|dup)     17   
6                      Remaining valid hgvs alleles     11   
7                      Remaining valid hgvs alleles      1   
8                      Remaining valid hgvs alleles      1   

               category  
0  Copy Number Variants  
1  Copy Number Variants  
2     Sequence Variants  
3     Sequence Variants  
4  Copy Number Variants  
5  Copy Number Variants  
6     Sequence Variants  
7     Sequence Variants  
8        Other Variants

In [52]:
clinvar_not_supported_df["category"] = clinvar_not_supported_df["in.variation_type"].map(clinvar_category_bins)
clinvar_not_supported_df.head(20)

Unnamed: 0          in.variation_type  in.vrs_xform_plan.policy  count  \
0            0           copy number gain       NCBI36 genomic only   2897   
1            1           copy number loss       NCBI36 genomic only   1749   
2            2                   Deletion  No hgvs or location info   1236   
3            3                  Diplotype        Genotype/Haplotype    596   
4            4                  Haplotype        Genotype/Haplotype    565   
5            5             Microsatellite  Invalid/unsupported hgvs    415   
6            6                   Deletion  Invalid/unsupported hgvs    413   
7            7                  Insertion  No hgvs or location info    382   
8            8                  Insertion  Invalid/unsupported hgvs    306   
9            9  single nucleotide variant  No hgvs or location info    287   
10          10              Translocation  No hgvs or location info    271   
11          11                Duplication  No hgvs or location info    265   
12          12       CompoundHeterozygote        Genotype/Haplotype    249   
13          13                  Variation  No hgvs or location info    206   
14          14                      Indel  No hgvs or location info    138   
15          15                Duplication  Invalid/unsupported hgvs    116   
16          16               protein only  No hgvs or location info     95   
17          17                    Complex  No hgvs or location info     72   
18          18                   Deletion       NCBI36 genomic only     71   
19          19             Microsatellite  No hgvs or location info     67   

                  category  
0     Copy Number Variants  
1     Copy Number Variants  
2        Sequence Variants  
3        Genotype Variants  
4        Sequence Variants  
5        Sequence Variants  
6        Sequence Variants  
7        Sequence Variants  
8        Sequence Variants  
9        Sequence Variants  
10  Rearrangement Variants  
11       Sequence Variants  
12       Genotype Variants  
13          Other Variants  
14       Sequence Variants  
15       Sequence Variants  
16       Sequence Variants  
17          Other Variants  
18       Sequence Variants  
19       Sequence Variants

Due to the structure of the data and the way that the original analysis developed, some but not all CNVs per the in.variation_type were annotated in the in.vrs_xform_plan.policy column as "Copy number change (cn loss|del and cn gain|dup)", "Absolute copy count", or "Min/max copy count range not supported".  However, some of the Copy number Gain/Loss variants did not get binned as CNVs per the in.vrs_xform_plan.policy.  Therefore, we need to mark those variants in the union of the following two sets as being in the category of Copy Number Variants:

Variants with in.variant_type ==
1. copy number loss
2. copy number gain

Variants with in.vrs_xform_plan.policy == 
1. Copy number change (cn loss|del and cn gain|dup)
2. Absolute copy count
3. Min/max copy count range not supported

Above we already caught the first set of variants. Now we must go back through each df one more time and map the variants we missed per in.vrs_xform_plan.policy values to the category of Copy Number Variants.

In [53]:
cnv_per_policy = ["Copy number change (cn loss|del and cn gain|dup)","Absolute copy count","Min/max copy count range not supported","Copy number change (cn loss|del and cn gain|dup)"]

In [54]:
clinvar_normalized_df.loc[
    clinvar_normalized_df["in.vrs_xform_plan.policy"].isin(cnv_per_policy),
      "category"
      ] = "Copy Number Variants"


In [55]:
clinvar_normalized_df

Unnamed: 0          in.variation_type  \
0            0  single nucleotide variant   
1            1                   Deletion   
2            2                Duplication   
3            3           copy number loss   
4            4             Microsatellite   
5            5           copy number gain   
6            6                   Deletion   
7            7                      Indel   
8            8                  Insertion   
9            9                Duplication   
10          10           copy number loss   
11          11                  Inversion   
12          12           copy number gain   
13          13  single nucleotide variant   
14          14                      Indel   
15          15                  Variation   
16          16                  Insertion   
17          17                   Deletion   
18          18                  Variation   
19          19                Duplication   

                            in.vrs_xform_plan.policy    count  \
0                                     Canonical SPDI  1934692   
1                                     Canonical SPDI    92875   
2                                     Canonical SPDI    43185   
3                                Absolute copy count    26857   
4                                     Canonical SPDI    26678   
5                                Absolute copy count    25583   
6   Copy number change (cn loss|del and cn gain|dup)    16249   
7                                     Canonical SPDI    10986   
8                                     Canonical SPDI     8865   
9   Copy number change (cn loss|del and cn gain|dup)     8119   
10  Copy number change (cn loss|del and cn gain|dup)     1524   
11                                    Canonical SPDI     1051   
12  Copy number change (cn loss|del and cn gain|dup)     1003   
13                      Remaining valid hgvs alleles      396   
14                      Remaining valid hgvs alleles      340   
15                                    Canonical SPDI      337   
16                      Remaining valid hgvs alleles      191   
17                               Absolute copy count        3   
18                      Remaining valid hgvs alleles        1   
19                               Absolute copy count        1   

                category  
0      Sequence Variants  
1      Sequence Variants  
2      Sequence Variants  
3   Copy Number Variants  
4      Sequence Variants  
5   Copy Number Variants  
6   Copy Number Variants  
7      Sequence Variants  
8      Sequence Variants  
9   Copy Number Variants  
10  Copy Number Variants  
11     Sequence Variants  
12  Copy Number Variants  
13     Sequence Variants  
14     Sequence Variants  
15        Other Variants  
16     Sequence Variants  
17  Copy Number Variants  
18        Other Variants  
19  Copy Number Variants

In [56]:
clinvar_not_normalized_df

Unnamed: 0          in.variation_type  \
0           0           copy number loss   
1           1           copy number gain   
2           2                   Deletion   
3           3                Duplication   
4           4           copy number loss   
5           5           copy number gain   
6           6  single nucleotide variant   
7           7                  Insertion   
8           8                  Variation   

                           in.vrs_xform_plan.policy  count  \
0                               Absolute copy count    475   
1                               Absolute copy count    344   
2  Copy number change (cn loss|del and cn gain|dup)     89   
3  Copy number change (cn loss|del and cn gain|dup)     72   
4  Copy number change (cn loss|del and cn gain|dup)     31   
5  Copy number change (cn loss|del and cn gain|dup)     17   
6                      Remaining valid hgvs alleles     11   
7                      Remaining valid hgvs alleles      1   
8                      Remaining valid hgvs alleles      1   

               category  
0  Copy Number Variants  
1  Copy Number Variants  
2     Sequence Variants  
3     Sequence Variants  
4  Copy Number Variants  
5  Copy Number Variants  
6     Sequence Variants  
7     Sequence Variants  
8        Other Variants

In [57]:
clinvar_not_normalized_df.loc[
    clinvar_not_normalized_df["in.vrs_xform_plan.policy"].isin(cnv_per_policy),
      "category"
      ] = "Copy Number Variants"


clinvar_not_normalized_df

Unnamed: 0          in.variation_type  \
0           0           copy number loss   
1           1           copy number gain   
2           2                   Deletion   
3           3                Duplication   
4           4           copy number loss   
5           5           copy number gain   
6           6  single nucleotide variant   
7           7                  Insertion   
8           8                  Variation   

                           in.vrs_xform_plan.policy  count  \
0                               Absolute copy count    475   
1                               Absolute copy count    344   
2  Copy number change (cn loss|del and cn gain|dup)     89   
3  Copy number change (cn loss|del and cn gain|dup)     72   
4  Copy number change (cn loss|del and cn gain|dup)     31   
5  Copy number change (cn loss|del and cn gain|dup)     17   
6                      Remaining valid hgvs alleles     11   
7                      Remaining valid hgvs alleles      1   
8                      Remaining valid hgvs alleles      1   

               category  
0  Copy Number Variants  
1  Copy Number Variants  
2  Copy Number Variants  
3  Copy Number Variants  
4  Copy Number Variants  
5  Copy Number Variants  
6     Sequence Variants  
7     Sequence Variants  
8        Other Variants

In [58]:
clinvar_not_supported_df

Unnamed: 0          in.variation_type  \
0            0           copy number gain   
1            1           copy number loss   
2            2                   Deletion   
3            3                  Diplotype   
4            4                  Haplotype   
5            5             Microsatellite   
6            6                   Deletion   
7            7                  Insertion   
8            8                  Insertion   
9            9  single nucleotide variant   
10          10              Translocation   
11          11                Duplication   
12          12       CompoundHeterozygote   
13          13                  Variation   
14          14                      Indel   
15          15                Duplication   
16          16               protein only   
17          17                    Complex   
18          18                   Deletion   
19          19             Microsatellite   
20          20                  Inversion   
21          21                Duplication   
22          22                      Indel   
23          23                  Inversion   
24          24  Haplotype, single variant   
25          25           copy number gain   
26          26              Phase unknown   
27          27           copy number loss   
28          28                     fusion   
29          29           copy number gain   
30          30                    Complex   
31          31                  Variation   
32          32                    Complex   
33          33              Translocation   
34          34              Translocation   
35          35         Tandem duplication   
36          36       Distinct chromosomes   

                  in.vrs_xform_plan.policy  count                category  
0                      NCBI36 genomic only   2897    Copy Number Variants  
1                      NCBI36 genomic only   1749    Copy Number Variants  
2                 No hgvs or location info   1236       Sequence Variants  
3                       Genotype/Haplotype    596       Genotype Variants  
4                       Genotype/Haplotype    565       Sequence Variants  
5                 Invalid/unsupported hgvs    415       Sequence Variants  
6                 Invalid/unsupported hgvs    413       Sequence Variants  
7                 No hgvs or location info    382       Sequence Variants  
8                 Invalid/unsupported hgvs    306       Sequence Variants  
9                 No hgvs or location info    287       Sequence Variants  
10                No hgvs or location info    271  Rearrangement Variants  
11                No hgvs or location info    265       Sequence Variants  
12                      Genotype/Haplotype    249       Genotype Variants  
13                No hgvs or location info    206          Other Variants  
14                No hgvs or location info    138       Sequence Variants  
15                Invalid/unsupported hgvs    116       Sequence Variants  
16                No hgvs or location info     95       Sequence Variants  
17                No hgvs or location info     72          Other Variants  
18                     NCBI36 genomic only     71       Sequence Variants  
19                No hgvs or location info     67       Sequence Variants  
20                No hgvs or location info     63       Sequence Variants  
21                     NCBI36 genomic only     50       Sequence Variants  
22                Invalid/unsupported hgvs     40       Sequence Variants  
23                Invalid/unsupported hgvs     29       Sequence Variants  
24                      Genotype/Haplotype     21       Sequence Variants  
25  Min/max copy count range not supported     14    Copy Number Variants  
26                      Genotype/Haplotype      8          Other Variants  
27                Invalid/unsupported hgvs      7    Copy Number Variants  
28                No hgvs or location info      6         Fusion Variants  
29           

In [59]:
clinvar_not_supported_df.loc[
    clinvar_not_supported_df["in.vrs_xform_plan.policy"].isin(cnv_per_policy),
      "category"
      ] = "Copy Number Variants"

clinvar_not_supported_df

Unnamed: 0          in.variation_type  \
0            0           copy number gain   
1            1           copy number loss   
2            2                   Deletion   
3            3                  Diplotype   
4            4                  Haplotype   
5            5             Microsatellite   
6            6                   Deletion   
7            7                  Insertion   
8            8                  Insertion   
9            9  single nucleotide variant   
10          10              Translocation   
11          11                Duplication   
12          12       CompoundHeterozygote   
13          13                  Variation   
14          14                      Indel   
15          15                Duplication   
16          16               protein only   
17          17                    Complex   
18          18                   Deletion   
19          19             Microsatellite   
20          20                  Inversion   
21          21                Duplication   
22          22                      Indel   
23          23                  Inversion   
24          24  Haplotype, single variant   
25          25           copy number gain   
26          26              Phase unknown   
27          27           copy number loss   
28          28                     fusion   
29          29           copy number gain   
30          30                    Complex   
31          31                  Variation   
32          32                    Complex   
33          33              Translocation   
34          34              Translocation   
35          35         Tandem duplication   
36          36       Distinct chromosomes   

                  in.vrs_xform_plan.policy  count                category  
0                      NCBI36 genomic only   2897    Copy Number Variants  
1                      NCBI36 genomic only   1749    Copy Number Variants  
2                 No hgvs or location info   1236       Sequence Variants  
3                       Genotype/Haplotype    596       Genotype Variants  
4                       Genotype/Haplotype    565       Sequence Variants  
5                 Invalid/unsupported hgvs    415       Sequence Variants  
6                 Invalid/unsupported hgvs    413       Sequence Variants  
7                 No hgvs or location info    382       Sequence Variants  
8                 Invalid/unsupported hgvs    306       Sequence Variants  
9                 No hgvs or location info    287       Sequence Variants  
10                No hgvs or location info    271  Rearrangement Variants  
11                No hgvs or location info    265       Sequence Variants  
12                      Genotype/Haplotype    249       Genotype Variants  
13                No hgvs or location info    206          Other Variants  
14                No hgvs or location info    138       Sequence Variants  
15                Invalid/unsupported hgvs    116       Sequence Variants  
16                No hgvs or location info     95       Sequence Variants  
17                No hgvs or location info     72          Other Variants  
18                     NCBI36 genomic only     71       Sequence Variants  
19                No hgvs or location info     67       Sequence Variants  
20                No hgvs or location info     63       Sequence Variants  
21                     NCBI36 genomic only     50       Sequence Variants  
22                Invalid/unsupported hgvs     40       Sequence Variants  
23                Invalid/unsupported hgvs     29       Sequence Variants  
24                      Genotype/Haplotype     21       Sequence Variants  
25  Min/max copy count range not supported     14    Copy Number Variants  
26                      Genotype/Haplotype      8          Other Variants  
27                Invalid/unsupported hgvs      7    Copy Number Variants  
28                No hgvs or location info      6         Fusion Variants  
29           

Get counts from the three dfs.

In [60]:
category_counts

{'Copy Number Variants': [89, 0, 51, 140, 0.0],
 'Epigenetic Modification': [0, 0, 14, 14, 0.0],
 'Expression Variants': [0, 0, 298, 298, 0.0],
 'Fusion Variants': [1, 0, 294, 295, 0.0],
 'Gene Function Variants': [0, 0, 91, 91, 0.0],
 'Genotype Variants': [0, 0, 16, 16, 0.0],
 'Other Variants': [1, 0, 88, 89, 0.0],
 'Rearrangement Variants': [0, 0, 151, 151, 0.0],
 'Region-Defined Variants': [28, 3, 129, 160, 0.0],
 'Sequence Variants': [1937, 77, 680, 2694, 0.0]}

In [61]:
def sum_clinvar_counts(dataframe: pd.DataFrame, support_status: int) -> None:
    """given a dataframe and whether that dataframe represents normalized, not_normalized, or not_supported variants, adds the counts of variants to dictionary of counts
    
    :param dataframe: counts of variants in clinvar with variant type information in dataframe format.
    :param support_status: an int flag to indicate if the variants in the dataframe are normalized (0), unable to be normalized (1), or unsupported (2) by the normalizer
    """
    for i in category_counts.keys():
        subdf = dataframe[dataframe["category"] == i]
        if len(subdf):
            category = i
            count = subdf["count"].sum()
            print(category, count)
            category_counts[category][support_status] += count
            category_counts[category][Fields.TOTAL_COUNT] += count


In [62]:
sum_clinvar_counts(clinvar_normalized_df,Fields.NORMALIZED_COUNT)

category_counts

Copy Number Variants 79339
Other Variants 338
Sequence Variants 2119259


{'Copy Number Variants': [79428, 0, 51, 79479, 0.0],
 'Epigenetic Modification': [0, 0, 14, 14, 0.0],
 'Expression Variants': [0, 0, 298, 298, 0.0],
 'Fusion Variants': [1, 0, 294, 295, 0.0],
 'Gene Function Variants': [0, 0, 91, 91, 0.0],
 'Genotype Variants': [0, 0, 16, 16, 0.0],
 'Other Variants': [339, 0, 88, 427, 0.0],
 'Rearrangement Variants': [0, 0, 151, 151, 0.0],
 'Region-Defined Variants': [28, 3, 129, 160, 0.0],
 'Sequence Variants': [2121196, 77, 680, 2121953, 0.0]}

In [63]:
sum_clinvar_counts(clinvar_not_normalized_df,Fields.UNABLE_TO_NORMALIZE_COUNT)

category_counts

Copy Number Variants 1028
Other Variants 1
Sequence Variants 12


{'Copy Number Variants': [79428, 1028, 51, 80507, 0.0],
 'Epigenetic Modification': [0, 0, 14, 14, 0.0],
 'Expression Variants': [0, 0, 298, 298, 0.0],
 'Fusion Variants': [1, 0, 294, 295, 0.0],
 'Gene Function Variants': [0, 0, 91, 91, 0.0],
 'Genotype Variants': [0, 0, 16, 16, 0.0],
 'Other Variants': [339, 1, 88, 428, 0.0],
 'Rearrangement Variants': [0, 0, 151, 151, 0.0],
 'Region-Defined Variants': [28, 3, 129, 160, 0.0],
 'Sequence Variants': [2121196, 89, 680, 2121965, 0.0]}

In [64]:
sum_clinvar_counts(clinvar_not_supported_df,Fields.UNSUPPORTED_COUNT)

category_counts

Copy Number Variants 4672
Fusion Variants 6
Genotype Variants 845
Other Variants 293
Rearrangement Variants 274
Sequence Variants 4560


{'Copy Number Variants': [79428, 1028, 4723, 85179, 0.0],
 'Epigenetic Modification': [0, 0, 14, 14, 0.0],
 'Expression Variants': [0, 0, 298, 298, 0.0],
 'Fusion Variants': [1, 0, 300, 301, 0.0],
 'Gene Function Variants': [0, 0, 91, 91, 0.0],
 'Genotype Variants': [0, 0, 861, 861, 0.0],
 'Other Variants': [339, 1, 381, 721, 0.0],
 'Rearrangement Variants': [0, 0, 425, 425, 0.0],
 'Region-Defined Variants': [28, 3, 129, 160, 0.0],
 'Sequence Variants': [2121196, 89, 5240, 2126525, 0.0]}

## <a id='toc1_6_'></a>[Computing Coverage](#toc0_)

For the purposes of making the table, computing the percent of all variants normalized in each category.

In [65]:
for i in category_counts.keys():
    normalized = category_counts[i][Fields.NORMALIZED_COUNT]
    total = category_counts[i][Fields.TOTAL_COUNT]
    percent_covered = normalized/total
    category_counts[i][Fields.PERCENT_NORMALIZED] = "%.4f" % percent_covered

category_counts
    

{'Copy Number Variants': [79428, 1028, 4723, 85179, '0.9325'],
 'Epigenetic Modification': [0, 0, 14, 14, '0.0000'],
 'Expression Variants': [0, 0, 298, 298, '0.0000'],
 'Fusion Variants': [1, 0, 300, 301, '0.0033'],
 'Gene Function Variants': [0, 0, 91, 91, '0.0000'],
 'Genotype Variants': [0, 0, 861, 861, '0.0000'],
 'Other Variants': [339, 1, 381, 721, '0.4702'],
 'Rearrangement Variants': [0, 0, 425, 425, '0.0000'],
 'Region-Defined Variants': [28, 3, 129, 160, '0.1750'],
 'Sequence Variants': [2121196, 89, 5240, 2126525, '0.9975']}

**Table of contents**<a id='toc0_'></a>    
- [Normalizer Performance Analysis](#toc1_)    
  - [Import relevant packages](#toc1_1_)    
  - [Dictionaries to map variants to categories and record category counts](#toc1_2_)    
  - [CIViC](#toc1_3_)    
  - [MOA](#toc1_4_)    
  - [ClinVar](#toc1_5_)    
  - [Computing Coverage](#toc1_6_)    
  - [Generating Table](#toc1_7_)    

<!-- vscode-jupyter-toc-config
	numbering=false
	anchor=true
	flat=false
	minLevel=1
	maxLevel=6
	/vscode-jupyter-toc-config -->
<!-- THIS CELL WILL BE REPLACED ON TOC UPDATE. DO NOT WRITE YOUR TEXT IN THIS CELL -->

## <a id='toc1_7_'></a>[Generating Table](#toc0_)

Generating a table in plotly to show variant counts and normalization percentage by category, as well as the types of data fields associated with different variant categories.

In [66]:
class VariantCategory(str, Enum):
    """Create enum for the kind of variants that are in the combined analysis."""
    SEQUENCE_VARS = "Sequence Variants"
    GENOTYPES = "Genotype Variants"
    FUSION = "Fusion Variants"
    REARRANGEMENTS = "Rearrangement Variants"
    EPIGENETIC_MODIFICATION = "Epigenetic Modification"
    COPY_NUMBER = "Copy Number Variants"
    EXPRESSION = "Expression Variants"
    GENE_FUNC = "Gene Function Variants"
    REGION_DEFINED_VAR = "Region-Defined Variants"
    OTHER = "Other Variants"

VARIANT_CATEGORY_VALUES = VariantCategory.__members__.values()
    

In [67]:
base_colors = ['rgb(49, 130, 189)','rgb(239, 243, 255)', 'rgb(189, 215, 231)', 'rgb(107, 174, 214)',
           'white']
core_field = "\u2B24"
optional_field = "<b>◯</b>"

colors = ['rgb(49, 130, 189)','white', 'white', 'white',
           'white', 'rgb(49, 130, 189)', 'white', 'white', 'rgb(189, 215, 231)','rgb(107, 174, 214)']

data = {
        # 'variant_category' : VARIANT_CATEGORY_VALUES,
        'variant_category' : ["Sequence", "Genotype", "Fusion", "Rearrangement", "Epigenetic Modification", "Copy Number", "Expression", "Gene Function", "Region-Defined", "Other"],
        'counts' : [f'{category_counts[v.value][Fields.TOTAL_COUNT]:,}' for v in VARIANT_CATEGORY_VALUES],
        'percent_normalized' : ["%.2f" %round(float(category_counts[v.value][Fields.PERCENT_NORMALIZED])*100,2)+"%" for v in VARIANT_CATEGORY_VALUES],
        'delta_sequence' : [core_field, core_field, "", "", "", "", "", "", "", optional_field],
        'delta_location' : [optional_field, optional_field, core_field, core_field, "", "", "", "", "", ""],
        'delta_frame' : [optional_field, optional_field, "", "", "", "", "", "", "", optional_field],
        'delta_quantity' : [optional_field, optional_field, "", "", core_field, core_field, optional_field, "", "", optional_field],
        'delta_function' : [optional_field, optional_field, "", "", optional_field, optional_field, core_field, core_field, "", optional_field],
        'region_specificity' : [optional_field, optional_field, optional_field, optional_field, optional_field, optional_field, optional_field, optional_field, core_field, optional_field],
        'shading' : colors
         }
df = pd.DataFrame(data)

fig = go.Figure(data=[go.Table(
  columnwidth = [90,53,65,53,50,50,50,50,50,50],
  header=dict(
    values=["<b>Variant Category</b>", "<b>Count</b>", "<b>% Normalized</b>", "<b>Δ Sequence</b>", "<b>Δ Location</b>", "<b>Δ Frame</b>", "<b>Δ Quantity</b>", "<b>Δ Function</b>", "<b>Region Specificity</b>"],
    line_color='black', fill_color='white',
    align='center', font=dict(color='black', size=18)
  ),
  cells=dict(
    values=[df.variant_category, df.counts, df.percent_normalized, df.delta_sequence, df.delta_location, df.delta_frame, df.delta_quantity, df.delta_function, df.region_specificity],
    line_color=["black"], fill_color= [df.shading],
    align='right', font=dict(color='black', size=18), height=30
  ))

])

fig.add_annotation(
            dict(
                text='  \u2B24  Core information fields<br><br>  <b>◯</b>  Optional information fields  ',
                align='left',
                showarrow=False,
                xref='paper',
                xanchor = 'right',
                yref='paper',
                x=0.98,
                y=0.02,
                yanchor = 'bottom',
                bordercolor='black',
                borderwidth=1
            ))

fig.update_layout(
    height=585, 
    width=1400,
    font=dict(
        size=18,
        color="Black"
        ),
    title = "<b>Counts, Normalizer Performance, and Data Types of Variants by Category</b>",
        margin=go.layout.Margin(
        l=2, #left margin
        r=2, #right margin
        b=0, #bottom margin
        t=52  #top margin
    ))
fig.show()

Exporting the table as a .png file.

In [68]:
fig.write_image("../merged_performance_analysis_table.png",'png')